Bluetooth: Controller: Cleanup mutual exclusion in CPR implementation

cvinayak opened the pull request:
Minor cleanup of the implementation of mutual exclusion
of LE Connection Parameter Request amongst active
connections.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Cleanup mutual exclusion in CPR implementation

Minor cleanup of the implementation of mutual exclusion
of LE Connection Parameter Request amongst active
connections.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

thoh-ot commented:
What about adding an small inline interface to make the CPR mutex logic even more clearer?
`void ull_conn_cpr_mutex_take(struct *ll_conn)`
`void ull_conn_cpr_mutex_give(struct *ll_conn)`
`struct *ll_conn ull_conn_cpr_mutex_owner_get(void)`

cvinayak commented:
> 
> 
> What about adding an small inline interface to make the CPR mutex logic even more clearer?
> `void ull_conn_cpr_mutex_take(struct *ll_conn)`
> `void ull_conn_cpr_mutex_give(struct *ll_conn)`
> `struct *ll_conn ull_conn_cpr_mutex_owner_get(void)`

Agree, will do the changes and update this PR when done.

cvinayak committed:
Bluetooth: Controller: Use inline functions for CPR mutex operations

Use inline functions for CPR mutex operations.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented:
> 
> 
> > What about adding an small inline interface to make the CPR mutex logic even more clearer?
> > `void ull_conn_cpr_mutex_take(struct *ll_conn)`
> > `void ull_conn_cpr_mutex_give(struct *ll_conn)`
> > `struct *ll_conn ull_conn_cpr_mutex_owner_get(void)`
> 
> Agree, will do the changes and update this PR when done.

@thoh-ot done. Please review.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_conn.c:
```diff
@@ -2437,6 +2433,43 @@ static bool is_enc_req_pause_tx(struct ll_conn *conn)
+#if defined(CONFIG_BT_CTLR_CONN_PARAM_REQ)
+/* Connection context pointer used as CPR mutex to serialize connection
+ * parameter requests procedures across simulataneous connections so that
+ * offsets exchanged to the peer does not get changed.
```
```suggestion
 * offsets exchanged to the peer do not get changed.
```

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_conn.c:
```diff
@@ -2437,6 +2433,43 @@ static bool is_enc_req_pause_tx(struct ll_conn *conn)
+ */
+static struct ll_conn *conn_upd_curr;
+
+static inline void cpr_mutex_check_and_take(struct ll_conn *conn)
+{
+	if (!conn_upd_curr) {
+		conn_upd_curr = conn;
+	}
+}
+
+static inline void cpr_mutex_take(struct ll_conn *conn)
+{
+	conn_upd_curr = conn;
+}
+
+static inline bool cpr_mutex_is_taken(struct ll_conn *conn)
+{
+	return conn_upd_curr && (conn_upd_curr != conn);
+}
+
+static inline void cpr_mutex_check_and_give(struct ll_conn *conn)
+{
+	if (conn == conn_upd_curr) {
+		conn_upd_curr = NULL;
+	}
+}
+
+static inline void cpr_mutex_give(void)
+{
+	conn_upd_curr = NULL;
+}
```
I'm not particularly fond of calling a simple static pointer value for a "mutex". I suggest that you either implement an actual thread-safe mutex, or remove the notation of "mutex" from these functions.

cvinayak replied:
This is a connection parameter request procedure mutual exclusion and nothing to do with OS mutex feature. Suggestions on naming are welcome.

@thoh-ot any other alternative to your initial proposal?

Thalley replied:
A simple renaming could be to just do replace `mutex` with `pointer`, and possibly rename `give` to `release`.

cvinayak replied:
Renamed to `cpr_active_set` and `cpr_active_reset`

cvinayak committed:
Bluetooth: Controller: Rename CPR mutual exclusion functions

Rename CPR mutual exclusion function to not sound similar to
OS mutex interfaces.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_conn.c:
```diff
@@ -113,6 +111,8 @@ static inline bool ctrl_is_unexpected(struct ll_conn *conn, uint8_t opcode);
 #endif /* CONFIG_BT_CTLR_LE_ENC */
 
 #if defined(CONFIG_BT_CTLR_CONN_PARAM_REQ)
+static inline void cpr_active_check_and_reset(struct ll_conn *conn);
+static inline void cpr_active_reset(void);
```
Is there any good reason why these are `inline`? The compiler should be able to optimize it well based on the optimization parameter. 

See also https://www.kernel.org/doc/html/v4.10/process/coding-style.html#the-inline-disease

The comment applies to all the new functions added.

cvinayak replied:
These functions operate on global variable (compile time constant), do nothing more that assignments. Since, the functions are called more than a single place, if compile optimizes for space (which is the default in zephyr), then there will be icache miss leading to wait states which is in addition to branching operation.

Thalley replied:
Interesting. Is this something you have seen/measured before? 

Either way, if that is the case, could I bother you to add that as a comment somewhere in the file, so that it is clear, and so that no one else attempts to remove the `inline` for these functions?

cvinayak replied:
Yes, seen and measured some of the inline function use in the controller. If I recollect, for example `adv_pdu_adva_get` in ull_adv.c has two callers and without `inline` compiler branches to the function.
